browser/ui/tabs/split_view_tab_strip_model_adapter.cc

```diff
@@ -0,0 +1,258 @@
+    std::swap(tab1, other_tab);
+    CHECK(tab1 == changed_tab_handle);
+  }
+
+  if (model_->IsTabPinned(index) ==
+      model_->IsTabPinned(model_->GetIndexOfTab(other_tab))) {
+    return;
+  }
+
+  base::SequencedTaskRunner::GetCurrentDefault()->PostTask(
+      FROM_HERE, base::BindOnce(
+                     [](base::WeakPtr<SplitViewTabStripModelAdapter> adapter,
+                        SplitViewBrowserData::Tile tile, tabs::TabHandle tab,
+                        bool pinned) {
+                       if (!adapter) {
+                         return;
+                       }
+
+                       adapter->model_->SetTabPinned(
+                           adapter->model_->GetIndexOfTab(tab), pinned);
+                       adapter->MakeTiledTabsAdjacent(tile, true);
+                     },
+                     weak_ptr_factory_.GetWeakPtr(), tile, other_tab,
+                     model_->IsTabPinned(index)));
+}
+
+void SplitViewTabStripModelAdapter::TabGroupedStateChanged(
+    std::optional<tab_groups::TabGroupId> group,
+    content::WebContents* contents,
+    int index) {
+  // In case a tiled tab is grouped or ungrouped, we need to synchronize the
+  // other tab together.
+  auto changed_tab_handle = model_->GetTabHandleAt(index);
+  if (!split_view_browser_data_->IsTabTiled(changed_tab_handle)) {
+    return;
+  }
+
+  auto tile = split_view_browser_data_->GetTile(changed_tab_handle);
+  auto [tab1, other_tab] = tile;
+  if (tab1 != changed_tab_handle) {
+    std::swap(tab1, other_tab);
+    CHECK(tab1 == changed_tab_handle);
+  }
+
+  auto other_tab_index = model_->GetIndexOfTab(other_tab);
+  auto tab_group_for_secondary_tab = model_->GetTabGroupForTab(other_tab_index);
+  if (group == tab_group_for_secondary_tab) {
+    return;
+  }
+
+  base::SequencedTaskRunner::GetCurrentDefault()->PostTask(
+      FROM_HERE,
+      base::BindOnce(
+          [](base::WeakPtr<SplitViewTabStripModelAdapter> adapter,
+             std::optional<tab_groups::TabGroupId> group_id,
+             SplitViewBrowserData::Tile tile, tabs::TabHandle tab_to_move) {
+            if (!adapter) {
+              return;
+            }
+
+            auto tab_index = adapter->model_->GetIndexOfTab(tab_to_move);
+            if (group_id.has_value()) {
+              adapter->model_->AddToExistingGroup({tab_index},
+                                                  group_id.value());
+            } else {
+              adapter->model_->RemoveFromGroup({tab_index});
+            }
+
+            adapter->MakeTiledTabsAdjacent(tile, true);
+          },
+          weak_ptr_factory_.GetWeakPtr(), group, tile, other_tab));
+}
+
+void SplitViewTabStripModelAdapter::OnTabRemoved(
+    const TabStripModelChange::Remove* remove) {
+  for (auto& removed_tabs : remove->contents) {
+    if (removed_tabs.remove_reason !=
+        TabStripModelChange::RemoveReason::kDeleted) {
+      // In case the remove_reason is kCached or kInsertedIntoOtherTabStrip,
```

Comment:
At this point I think we've already called `BreakTile` in `OnTabWillBeRemoved` - or is that only triggered when the user closes a tab?

Reply:
Yeah we already broke that, but we should 
* restore the tile when the kCached tabs are restore inside OnTabInserted(). 
* transfer the boken tile to the other SplitViewBrowserData that's bound to the new browser with the new tabstrip.

I have yet to verify the sequence, so they are kind of todos